coveo-functools/tests_functools/test_flex_adapters.py

```diff
@@ -130,3 +132,25 @@ def abstract_adapter(value: Any) -> TypeHint:
     assert "test" in payload
     assert isinstance(instance.test, Implementation)
     assert instance.test.value == "success"
+
+
+@dataclass
+class TestFactory:
+    value: str
+
+    @classmethod
+    def factory(cls, raw: Dict[str, str]) -> TestFactory:
+        return deserialize(raw, hint=TestFactory)
+
+
+@parametrize('raw', (True, False))
+def test_deserialize_adapter_factory(raw: bool) -> None:
+    """The adapter may return a callable."""
+
+    def factory_adapter(value: Any) -> TypeHint:
+        return TestFactory.factory if 'raw' in value else TestFactory
+
+    register_subclass_adapter(TestFactory, factory_adapter)
+
+    payload: Dict[str, Any] = {'raw': {"value": "success"}} if raw else {"value": "success"}
+    assert deserialize(payload, hint=TestFactory).value == 'success'

```

Comment:
This is not a huge deal, but I think a lot of the work you do in the test is there to make it work with parameterization. Hence, I'd think this test is not a good candidate to parameterization. I think it'd be clearer if you just write a different test explicitly for each case.